File: backend/services/harvest_service.go
package services
import (
	"errors"
	"fmt"
	"time"
	"cronos-backend/blockchain"
	"cronos-backend/models"
	"cronos-backend/repositories"
	"github.com/google/uuid"
)
type HarvestService struct {
	HarvestRepo *repositories.HarvestRepository
	Ledger      *blockchain.Ledger
}
func NewHarvestService(harvestRepo *repositories.HarvestRepository, ledger *blockchain.Ledger) *HarvestService {
	return &HarvestService{HarvestRepo: harvestRepo, Ledger: ledger}
}
func (s *HarvestService) CreateHarvest(userID uuid.UUID, harvest *models.Harvest) error {
	if err := s.HarvestRepo.CreateHarvest(harvest); err != nil {
		return errors.New("failed to create harvest")
	}
	s.Ledger.RecordEvent("harvest_completed", userID, "harvest", harvest.ID, map[string]interface{}{
		"total_weight":  harvest.TotalWeight,
		"shrimp_size":   harvest.ShrimpSize,
		"quality_grade": harvest.QualityGrade,
	})
	return nil
}
func (s *HarvestService) GetMyHarvests(userID uuid.UUID) ([]models.Harvest, error) {
	return s.HarvestRepo.FindHarvestsByUserID(userID)
}
func (s *HarvestService) GetHarvest(id uuid.UUID) (*models.Harvest, error) {
	return s.HarvestRepo.FindHarvestByID(id)
}
func (s *HarvestService) CreateBatch(userID uuid.UUID, batch *models.Batch) error {
	serial, err := s.HarvestRepo.GetNextBatchSerial()
	if err != nil {
		return errors.New("failed to generate batch code")
	}
	year := time.Now().Year()
	batch.BatchCode = fmt.Sprintf("CRN-VNM-%d-%06d", year, serial)
	if err := s.HarvestRepo.CreateBatch(batch); err != nil {
		return errors.New("failed to create batch")
	}
	s.Ledger.RecordEvent("batch_created", userID, "batch", batch.ID, map[string]interface{}{
		"batch_code": batch.BatchCode,
		"quantity":   batch.Quantity,
	})
	return nil
}
func (s *HarvestService) GetMyBatches(userID uuid.UUID) ([]models.Batch, error) {
	return s.HarvestRepo.FindBatchesByUserID(userID)
}
func (s *HarvestService) GetBatchByCode(code string) (*models.Batch, error) {
	return s.HarvestRepo.FindBatchByCode(code)
}

File: backend-spring/src/main/java/com/cronos/service/HarvestService.java
package com.cronos.service;

import com.cronos.blockchain.BlockchainLedger;
import com.cronos.entity.Batch;
import com.cronos.entity.Harvest;
import com.cronos.repository.BatchRepository;
import com.cronos.repository.HarvestRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import java.time.Year;
import java.util.*;

@Service
@RequiredArgsConstructor
public class HarvestService {
    private final HarvestRepository harvestRepo;
    private final BatchRepository batchRepo;
    private final BlockchainLedger ledger;

    public Harvest createHarvest(UUID userId, Harvest harvest) {
        harvestRepo.save(harvest);
        ledger.recordEvent("harvest_completed", userId, "harvest", harvest.getId(),
                Map.of("total_weight", harvest.getTotalWeight(), "shrimp_size", harvest.getShrimpSize() != null ? harvest.getShrimpSize() : "",
                        "quality_grade", harvest.getQualityGrade() != null ? harvest.getQualityGrade() : ""));
        return harvest;
    }

    public List<Harvest> getMyHarvests(UUID userId) { return harvestRepo.findByUserId(userId); }
    public Harvest getHarvest(UUID id) { return harvestRepo.findById(id).orElseThrow(() -> new RuntimeException("harvest not found")); }

    public Batch createBatch(UUID userId, Batch batch) {
        long serial = batchRepo.count() + 1;
        batch.setBatchCode(String.format("CRN-VNM-%d-%06d", Year.now().getValue(), serial));
        batchRepo.save(batch);
        ledger.recordEvent("batch_created", userId, "batch", batch.getId(),
                Map.of("batch_code", batch.getBatchCode(), "quantity", batch.getQuantity()));
        return batch;
    }

    public List<Batch> getMyBatches(UUID userId) { return batchRepo.findByUserId(userId); }
    public Batch getBatchByCode(String code) { return batchRepo.findByBatchCode(code).orElseThrow(() -> new RuntimeException("batch not found")); }
}
